File: src/browser_stream_protocol.cpp
/**
 * @file src/browser_stream_protocol.cpp
 * @brief Browser Stream media envelope helpers.
 */

#include "browser_stream_protocol.h"

#include <algorithm>

namespace browser_stream::protocol {
  namespace {
    constexpr std::uint32_t magic = 0x31534250;  // "PBS1" little-endian.

    void write_u16(std::vector<std::uint8_t> &out, std::size_t offset, std::uint16_t value) {
      out[offset] = static_cast<std::uint8_t>(value & 0xff);
      out[offset + 1] = static_cast<std::uint8_t>((value >> 8) & 0xff);
    }

    void write_u32(std::vector<std::uint8_t> &out, std::size_t offset, std::uint32_t value) {
      for (int i = 0; i < 4; ++i) {
        out[offset + i] = static_cast<std::uint8_t>((value >> (8 * i)) & 0xff);
      }
    }

    void write_u64(std::vector<std::uint8_t> &out, std::size_t offset, std::uint64_t value) {
      for (int i = 0; i < 8; ++i) {
        out[offset + i] = static_cast<std::uint8_t>((value >> (8 * i)) & 0xff);
      }
    }

    std::uint16_t read_u16(std::span<const std::uint8_t> data, std::size_t offset) {
      return static_cast<std::uint16_t>(data[offset]) |
             static_cast<std::uint16_t>(data[offset + 1] << 8);
    }

    std::uint32_t read_u32(std::span<const std::uint8_t> data, std::size_t offset) {
      std::uint32_t value = 0;
      for (int i = 0; i < 4; ++i) {
        value |= static_cast<std::uint32_t>(data[offset + i]) << (8 * i);
      }
      return value;
    }

    std::uint64_t read_u64(std::span<const std::uint8_t> data, std::size_t offset) {
      std::uint64_t value = 0;
      for (int i = 0; i < 8; ++i) {
        value |= static_cast<std::uint64_t>(data[offset + i]) << (8 * i);
      }
      return value;
    }
  }  // namespace

  std::vector<std::uint8_t> encode_datagram(const media_envelope_t &envelope) {
    std::vector<std::uint8_t> out(header_size + envelope.payload.size());
    write_u32(out, 0, magic);
    out[4] = version;
    out[5] = static_cast<std::uint8_t>(envelope.kind);
    out[6] = static_cast<std::uint8_t>(envelope.codec);
    out[7] = envelope.flags;
    write_u32(out, 8, envelope.sequence);
    write_u64(out, 12, envelope.frame_id);
    write_u64(out, 20, envelope.timestamp_us);
    write_u16(out, 28, envelope.chunk_index);
    write_u16(out, 30, envelope.chunk_count);
    write_u32(out, 32, static_cast<std::uint32_t>(envelope.payload.size()));
    std::copy(envelope.payload.begin(), envelope.payload.end(), out.begin() + header_size);
    return out;
  }

  bool decode_datagram(std::span<const std::uint8_t> datagram, media_envelope_t &envelope, std::string &error) {
    if (datagram.size() < header_size) {
      error = "Browser Stream datagram is shorter than the media header";
      return false;
    }

    if (read_u32(datagram, 0) != magic) {
      error = "Browser Stream datagram has an invalid magic value";
      return false;
    }

    if (datagram[4] != version) {
      error = "Browser Stream datagram has an unsupported version";
      return false;
    }

    const auto payload_size = read_u32(datagram, 32);
    if (payload_size != datagram.size() - header_size) {
      error = "Browser Stream datagram payload length does not match the header";
      return false;
    }

    const auto chunk_index = read_u16(datagram, 28);
    const auto chunk_count = read_u16(datagram, 30);
    if (chunk_count == 0 || chunk_index >= chunk_count) {
      error = "Browser Stream datagram has invalid chunk indexes";
      return false;
    }

    envelope.kind = static_cast<media_kind>(datagram[5]);
    envelope.codec = static_cast<codec_id>(datagram[6]);
    envelope.flags = datagram[7];
    envelope.sequence = read_u32(datagram, 8);
    envelope.frame_id = read_u64(datagram, 12);
    envelope.timestamp_us = read_u64(datagram, 20);
    envelope.chunk_index = chunk_index;
    envelope.chunk_count = chunk_count;
    envelope.payload.assign(datagram.begin() + header_size, datagram.end());
    error.clear();
    return true;
  }
}  // namespace browser_stream::protocol

File: src_assets/common/assets/web/browser-stream-protocol.js
export const BROWSER_STREAM_MAGIC = 0x31534250
export const BROWSER_STREAM_VERSION = 1
export const BROWSER_STREAM_HEADER_SIZE = 36

export const BrowserStreamKind = Object.freeze({
  video: 1,
  audio: 2,
})

export const BrowserStreamCodec = Object.freeze({
  h264: 1,
  opus: 2,
})

export const BrowserStreamFlags = Object.freeze({
  keyframe: 1,
})

export function encodeBrowserStreamControlFrame(message) {
  const encoder = new TextEncoder()
  const payload = encoder.encode(JSON.stringify(message))
  const out = new Uint8Array(4 + payload.byteLength)
  const view = new DataView(out.buffer)
  view.setUint32(0, payload.byteLength, false)
  out.set(payload, 4)
  return out
}

export function encodeBrowserStreamDatagram(envelope) {
  const payload = envelope.payload instanceof Uint8Array ? envelope.payload : new Uint8Array(envelope.payload || [])
  const out = new Uint8Array(BROWSER_STREAM_HEADER_SIZE + payload.byteLength)
  const view = new DataView(out.buffer)

  view.setUint32(0, BROWSER_STREAM_MAGIC, true)
  view.setUint8(4, BROWSER_STREAM_VERSION)
  view.setUint8(5, envelope.kind)
  view.setUint8(6, envelope.codec)
  view.setUint8(7, envelope.flags || 0)
  view.setUint32(8, envelope.sequence || 0, true)
  view.setBigUint64(12, BigInt(envelope.frameId || 0), true)
  view.setBigUint64(20, BigInt(envelope.timestampUs || 0), true)
  view.setUint16(28, envelope.chunkIndex || 0, true)
  view.setUint16(30, envelope.chunkCount || 1, true)
  view.setUint32(32, payload.byteLength, true)
  out.set(payload, BROWSER_STREAM_HEADER_SIZE)
  return out
}

export function decodeBrowserStreamDatagram(datagram) {
  const bytes = datagram instanceof Uint8Array ? datagram : new Uint8Array(datagram)
  if (bytes.byteLength < BROWSER_STREAM_HEADER_SIZE) {
    throw new Error('Browser Stream datagram is shorter than the media header')
  }

  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
  if (view.getUint32(0, true) !== BROWSER_STREAM_MAGIC) {
    throw new Error('Browser Stream datagram has an invalid magic value')
  }
  if (view.getUint8(4) !== BROWSER_STREAM_VERSION) {
    throw new Error('Browser Stream datagram has an unsupported version')
  }

  const payloadLength = view.getUint32(32, true)
  if (payloadLength !== bytes.byteLength - BROWSER_STREAM_HEADER_SIZE) {
    throw new Error('Browser Stream datagram payload length does not match the header')
  }

  const chunkIndex = view.getUint16(28, true)
  const chunkCount = view.getUint16(30, true)
  if (chunkCount === 0 || chunkIndex >= chunkCount) {
    throw new Error('Browser Stream datagram has invalid chunk indexes')
  }

  return {
    kind: view.getUint8(5),
    codec: view.getUint8(6),
    flags: view.getUint8(7),
    sequence: view.getUint32(8, true),
    frameId: Number(view.getBigUint64(12, true)),
    timestampUs: Number(view.getBigUint64(20, true)),
    chunkIndex,
    chunkCount,
    payload: bytes.slice(BROWSER_STREAM_HEADER_SIZE),
  }
}

export function certHashHexToBytes(hex) {
  const clean = (hex || '').replace(/[^a-fA-F0-9]/g, '')
  if (clean.length % 2 !== 0) {
    throw new Error('Certificate hash must contain an even number of hex digits')
  }
  const bytes = new Uint8Array(clean.length / 2)
  for (let i = 0; i < clean.length; i += 2) {
    bytes[i / 2] = Number.parseInt(clean.slice(i, i + 2), 16)
  }
  return bytes
}

export function createBrowserStreamFrameAssembler(options = {}) {
  const maxFrameAgeMs = options.maxFrameAgeMs ?? 250
  const maxPendingFrames = options.maxPendingFrames ?? 8
  const pending = new Map()

  function pendingKey(datagram) {
    return `${datagram.kind}:${datagram.codec}:${datagram.frameId}`
  }

  function prune(nowMs, newestFrameId) {
    let droppedFrames = 0
    for (const [key, frame] of pending) {
      if (nowMs - frame.firstSeenMs > maxFrameAgeMs || frame.frameId + maxPendingFrames < newestFrameId) {
        pending.delete(key)
        droppedFrames += 1
      }
    }

    while (pending.size > maxPendingFrames) {
      const key = pending.keys().next().value
      pending.delete(key)
      droppedFrames += 1
    }
    return droppedFrames
  }

  function assemble(frame) {
    const payload = new Uint8Array(frame.receivedBytes)
    let offset = 0
    for (const chunk of frame.chunks) {
      payload.set(chunk, offset)
      offset += chunk.byteLength
    }
    return {
      kind: frame.kind,
      codec: frame.codec,
      flags: frame.flags,
      frameId: frame.frameId,
      timestampUs: frame.timestampUs,
      payload,
    }
  }

  return {
    get pendingCount() {
      return pending.size
    },
    reset() {
      pending.clear()
    },
    push(datagram, nowMs = performance.now()) {
      let droppedFrames = prune(nowMs, datagram.frameId)

      if (datagram.chunkCount === 1) {
        return {
          frame: {
            kind: datagram.kind,
            codec: datagram.codec,
            flags: datagram.flags,
            frameId: datagram.frameId,
            timestampUs: datagram.timestampUs,
            payload: datagram.payload,
          },
          droppedFrames,
        }
      }

      const key = pendingKey(datagram)
      let frame = pending.get(key)
      if (!frame) {
        frame = {
          kind: datagram.kind,
          codec: datagram.codec,
          flags: datagram.flags,
          frameId: datagram.frameId,
          timestampUs: datagram.timestampUs,
          chunkCount: datagram.chunkCount,
          chunks: new Array(datagram.chunkCount),
          receivedChunks: 0,
          receivedBytes: 0,
          firstSeenMs: nowMs,
        }
        pending.set(key, frame)
      }

      if (frame.chunkCount !== datagram.chunkCount || frame.chunks[datagram.chunkIndex]) {
        return { frame: null, droppedFrames }
      }

      frame.chunks[datagram.chunkIndex] = datagram.payload
      frame.receivedChunks += 1
      frame.receivedBytes += datagram.payload.byteLength

      if (frame.receivedChunks !== frame.chunkCount) {
        return { frame: null, droppedFrames }
      }

      pending.delete(key)
      return { frame: assemble(frame), droppedFrames }
    },
  }
}
